### PR TITLE
Fix: Stop Construction Windows Closing on Company Rename (Issue #7479)

### PR DESCRIPTION
**Pull request by bendevnull**
While unaware of the ramifications or importance of this "fix", I completed it anyway as per request of issue #7479. If this is still a bad idea, please let me know why and I will try to find a compromise for both you and the issue author. Thanks :)

### CODE REVIEW

**Comment by nielsmh** [comment]
I think the risk is invalid windows remaining open if switching to another company, or even switching to being a spectator. You should probably add a variable testing if the new company ID is equal to the old and only close windows if they are different.

**Comment by bendevnull** [comment]
Or better yet, invalidate the windows to recreate them :sunglasses:

File: src/company_cmd.cpp [hunk 1]
```diff
@@ -108,11 +108,11 @@ void SetLocalCompany(CompanyID new_company)
 	InvalidateWindowData(WC_SEND_NETWORK_MSG, DESTTYPE_TEAM, _local_company);
 
 	assert(IsLocalCompany());
-
+
```
Comment (nielsmh) [inline src/company_cmd.cpp]: Blank lines should be completely blank, no trailing whitespace.

**Review by nielsmh — DISMISSED** [review]
Your changes are causing the regression tests to crash. I haven't investigated in detail, but check the logs from the automated builds.

**Comment by glx22** [comment]
Crashing during regression
```
+Crash reason:
+ERROR: Signal:  Segmentation fault (11)
+ERROR: Message: <none>
+
+OpenTTD version:
+ERROR: Version:    20190419--g105c051272 (0)
+ERROR: NewGRF ver: 1a006d64
+ERROR: Bits:       32
+ERROR: Endian:     little
+ERROR: Dedicated:  no
+ERROR: Build date: Apr 19 2019 00:35:11
+
+Stacktrace:
+ERROR: [00] ./openttd(_ZNK12CrashLogUnix13LogStacktraceEPcPKc+0x40) [0x569c0350]
+ERROR: [01] ./openttd(_ZNK8CrashLog12FillCrashLogEPcPKc+0xde) [0x56868a2e]
+ERROR: [02] ./openttd(_ZNK8CrashLog12MakeCrashLogEv+0x67) [0x56868cc7]
+ERROR: [03] ./openttd(+0x3ac27b) [0x569c027b]
+ERROR: [04] linux-gate.so.1(__kernel_sigreturn+0) [0xf7f4ffe0]
+ERROR: [05] ./openttd(_Z24InvalidateCompanyWindowsPK7Company+0x14) [0x56851104]
+ERROR: [06] ./openttd(_Z15SetLocalCompany5Owner+0x7c) [0x5685160c]
+ERROR: [07] ./openttd(_Z13GenerateWorld12GenWorldModejjb+0x91) [0x568c33b1]
+ERROR: [08] ./openttd(_Z12openttd_mainiPPc+0x1afc) [0x569aecfc]
+ERROR: [09] ./openttd(main+0x73) [0x5676d733]
+ERROR: [10] /lib/i386-linux-gnu/libc.so.6(__libc_start_main+0xf6) [0xf5c45286]
+ERROR: [11] ./openttd(+0x17cd25) [0x56790d25]

```

Edit: `GenerateWorld()` calls `SetLocalCompany(COMPANY_SPECTATOR)`, `Company::GetIfValid(new_company)` then returns `nullptr` and `InvalidateCompanyWindow(nullptr)` crashes

File: src/company_cmd.cpp [hunk 2]
```diff
@@ -108,11 +108,11 @@ void SetLocalCompany(CompanyID new_company)
...
 
 	assert(IsLocalCompany());
-
+	
 	_current_company = _local_company = new_company;
 
-	/* Delete any construction windows... */
-	DeleteConstructionWindows();
+	/* Recreate company windows */
+	InvalidateCompanyWindows(Company::GetIfValid(new_company));
```
Comment (glx22) [inline src/company_cmd.cpp]: `Company::GetIfValid(new_company)` can return `nullptr`, causing `company->index` to crash in `InvalidateCompanyWindow()`

Comment (bennyman123abc) [reply src/company_cmd.cpp]: Can Company::Get() return a `nullptr`?

**Comment by nielsmh** [comment]
I think you'll need to squash and force-push your branch for the commit checker to be happy.

**Comment by bendevnull** [comment]
Yep; the commit checker is satisfied now. :)

**Comment by bennyman123abc** [comment]
Bump?

File: src/company_cmd.cpp [hunk 3]
```diff
@@ -111,8 +111,11 @@ void SetLocalCompany(CompanyID new_company)
 
 	_current_company = _local_company = new_company;
 
-	/* Delete any construction windows... */
-	DeleteConstructionWindows();
+	/* Recreate company windows */
+	if (Company::IsValidID(new_company))
+	{
```
Comment (nielsmh) [inline src/company_cmd.cpp:116]: The opening brace of if, for, while etc. statements needs to go on the same line as the statement. Or make a braces-free single-statement body on one line.

**Review by nielsmh — CHANGES_REQUESTED** [review]
This fix is wrong. If you switch to Spectator in multiplayer then your construction windows don't close as they should.

The bug report is that construction windows close when you change your client name but remain in the same company. Hence the fix should be to only close the construction windows if the active company actually changed.

**Comment by LordAro** [comment]
Hi, there's been no activity on this for nearly 4 months. If no more effort will be put into this PR, it will be closed in the next week or so

**Comment by bendevnull** [comment]
This won't get finished. If someone else wants to take this over, feel free. Otherwise, I'm closing for now.

**Pull request closed without merging**